stan/math/prim/fun/quad_form.hpp

```diff
@@ -13,21 +13,21 @@ namespace math {
  * Symmetry of the resulting matrix is not guaranteed due to numerical
  * precision.
  *
- * @tparam RA number of rows in the square matrix, can be Eigen::Dynamic
- * @tparam CA number of columns in the square matrix, can be Eigen::Dynamic
- * @tparam RB number of rows in the second matrix, can be Eigen::Dynamic
- * @tparam CB number of columns in the second matrix, can be Eigen::Dynamic
- * @tparam T type of elements
+ * @tparam EigMat1 type of the first (square) matrix
+ * @tparam EigMat2 type of the second matrix
  *
  * @param A square matrix
  * @param B second matrix
  * @return The quadratic form, which is a symmetric matrix of size CB.
  * @throws std::invalid_argument if A is not square, or if A cannot be
  * multiplied by B
  */
-template <int RA, int CA, int RB, int CB, typename T>
-inline Eigen::Matrix<T, CB, CB> quad_form(const Eigen::Matrix<T, RA, CA>& A,
-                                          const Eigen::Matrix<T, RB, CB>& B) {
+template <typename EigMat1, typename EigMat2,
+          require_all_eigen_t<EigMat1, EigMat2>* = nullptr,
+          require_not_eigen_col_vector_t<EigMat2>* = nullptr,
+          require_vt_same<EigMat1, EigMat2>* = nullptr,
+          require_all_vt_arithmetic<EigMat1, EigMat2>* = nullptr>
```

Comment:
I'm guessing this is because Eigen doesn't implicitly promote, so mixed arithmetic value types wouldn't work?

Reply:
Exactley. If we really wanted to add support for mixed arithmetic we could `.cast<>()` both matrices to a common scalar type.

Reply:
Makes sense, thanks